switch_mod/project/unitcommit/commit.py

```diff
@@ -249,6 +255,72 @@ def define_components(mod):
+    mod.PROJ_MIN_UPTIME_DISPATCH_POINTS = Set(dimen=2, initialize=lambda m: [
+        (pr, tp) 
+            for pr in m.PROJECTS if m.g_min_uptime[m.proj_gen_tech[pr]] > 0.0
+                for tp in m.PROJ_ACTIVE_TIMEPOINTS[pr] 
+    ])
+    mod.PROJ_MIN_DOWNTIME_DISPATCH_POINTS = Set(dimen=2, initialize=lambda m: [
+        (pr, tp) 
+            for pr in m.PROJECTS if m.g_min_downtime[m.proj_gen_tech[pr]] > 0.0
+                for tp in m.PROJ_ACTIVE_TIMEPOINTS[pr] 
+    ])
+    
+    def tp_prev(m, tp, n=1):
+        # find nth previous timepoint, wrapping from start to end of day
+        return m.TS_TPS[m.tp_ts[tp]].prevw(tp, n)
+    # min_time_projects = set()
+    def min_time_rule(m, pr, tp, up):
+        # if (pr, up) not in min_time_projects:
+        #     min_time_projects.add((pr, up))
+        #     print "creating min {}time constraint for {}".format('up' if up else 'down', pr)
+        # note: this enforces a simple rule: all capacity turned on in the last x hours must 
+        # still be on now (or all capacity recently turned off must still be off).
+        # If there is a slice of capacity that is supposed to be forced out of service all day, 
+        # this could assume that some other capacity was turned off in the last x hours, but then 
+        # it is now turning on part of the forced-off capacity.
+        # This could be resolved by adding a term reflecting the minimum forced-off capacity over
+        # the last x hours, or by modeling each unit separately (so capacity is either 100% 
+        # available or 100% unavailable).
+
+        # how many timepoints must the project stay on/off once it's started/shutdown?
+        # note: we round to the nearest integer, so a project will be off for 1 timepoint if 
+        # g_min_downtime is 4 and ts_duration_of_tp is 3; if more conservative behavior is needed,
+        # the user should raise g_min_downtime to the desired multiple of ts_duration_of_tp
+        n_tp = int(round(m.g_min_downtime[m.proj_gen_tech[pr]] / m.ts_duration_of_tp[m.tp_ts[tp]]))
+        if n_tp == 1:
+            # project can be shutdown and restarted in the same timepoint
+            rule = Constraint.Skip
+        else:
+            # note: this rule stops one short of n_tp steps back (normal behavior of range()), because 
+            # the current timepoint is included in the duration when the capacity will be on/off.
+            if up:
+                rule = (    # online capacity >= recent startups (all recent startups are still online)
+                    m.CommitProject[pr, tp] 
+                    >= sum(m.Startup[pr, tp_prev(m, tp, i)] for i in range(1, n_tp))
+                )
+            else:
+                rule = (    # offline capacity >= recent shutdowns (all recent shutdowns are still offline)
+                    m.ProjCapacityTP[pr, tp] * m.proj_availability[pr] - m.CommitProject[pr, tp] 
```

Comment:
Use CommitSlackUp instead of ```m.ProjCapacityTP[pr, tp] * m.proj_availability[pr] - m.CommitProject[pr, tp]```. You forgot to account for proj_max_commit_fraction which is part of the upper bound of CommitProject, but it's included in CommitSlackUp.